### PR TITLE
Attach launched processes to systemd scopes

### PR DESCRIPTION
**Pull request by skewballfox**
- Depends on #53 
- Depends on [lauch-pad#12](https://github.com/pop-os/launch-pad/pull/12)
- Related to [cosmic-epoch#356](https://github.com/pop-os/cosmic-epoch/issues/356)

Still a draft, open to making any changes or taking an alternative approach. This adds a function for checking whether systemd is the init system ( I need to confirm the path linked is correct for all distros), and uses zbus to start a transient service for each process. 

I need to test this locally, and I want to implement a few other relevant methods from systemd. 

A relevant crate that might be worth considering as a dependency is [zbus_systemd](https://docs.rs/zbus_systemd). I wasn't if their version of start_transient_unit allowed for setting the PIDs property, and didn't want to add a dependency without asking in any case. It makes sense though if we wind up implementing more than a handful of methods.

### CODE REVIEW

**Comment by ryanabx** [comment]
I think it would make sense to include zbus_systemd as a dependency, behind a systemd feature, enabled by default

**Comment by skewballfox** [comment]
> I think it would make sense to include zbus_systemd as a dependency, behind a systemd feature, enabled by default

done. Do you think it would make sense to just feature gate the entire systemd file, along with the systemd calls in main?

**Comment by ryanabx** [comment]
> > I think it would make sense to include zbus_systemd as a dependency, behind a systemd feature, enabled by default
> 
> 
> 
> done. Do you think it would make sense to just feature gate the entire systemd file, along with the systemd calls in main? 

If they can be cleanly separated, yeah, that would be great!

**Comment by skewballfox** [comment]
should the started transient units have mode [fail or replace](https://www.freedesktop.org/software/systemd/man/latest/org.freedesktop.systemd1.html#:~:text=If%20%22replace%22%2C%20the%20method)? the modes are the same for transient unit. 

Right now I think they will fail if the scope name is not unique (may happen if user logs out and logs in, or cosmic-session restarts everything) so I think the correct mode is probably replace.

**Comment by skewballfox** [comment]
> If they can be cleanly separated, yeah, that would be great!

Depends on your definition of clean. there are 3 different calls to functions in systemd in main, I would probably wrap need to wrap each of them in a cfg attribute like this
```
#[cfg(feature = "systemd")]
	{
		if *is_systemd_used() {
	 ....
		}
	}
```
I think that's generally considered a bad practice in larger projects due to increased compile time, but I'm not sure it would make a noticeable difference here. Right now I just feature gated the code for spawning scopes. I'm fine with either option.

**Comment by skewballfox** [comment]
Since we are using dbus calls to spawn transient threads, do you think it would make sense to convert the start and stop systemd target stuff to use the systemd dbus interface? 

Also, do we need to set set the sytemd environment variable within cosmic-session, when the environment variables are passed to systemd from within [start-cosmic](https://github.com/pop-os/cosmic-session/blob/23c062951c2eb50ed0f88327501d0756ae9d912a/data/start-cosmic#L41). The one call to that function is setting `XDG_SESSION_TYPE` which is already passed by start-cosmic

File: src/systemd.rs
Comment (Drakulix) [inline src/systemd.rs:41]: Wouldn't it be enough to simply check, if the dbus interface exists? In other words, just don't treat an error of `spawn_scope` as fatal. Log it and continue operation.

File: src/main.rs
Comment (Drakulix) [inline src/main.rs:422]: So does this crash, if we don't have a pid?

File: src/main.rs
Comment (Drakulix) [inline src/main.rs:431]: Imo, we should just log this error and continue.

File: src/main.rs
Comment (skewballfox) [reply src/main.rs:422]: It wouldn't crash in that case, just spawn a scope that manages nothing. There is another systemd function to set a PIDs for a scope after the fact. 

I'm not sure there is a situation where it would be `None`. I made the value optional for two reasons. The first being, it's optional for [tokio child](https://docs.rs/tokio/latest/tokio/process/struct.Child.html#method.id), the second being I had to set it after processData has been created. 

There might be a situation where the PID would be incorrect though. If a child is spawned, and immedately restarts, I don't think it would reuse the same pid.

File: src/systemd.rs
Comment (skewballfox) [reply src/systemd.rs:41]: I do think we should have some sort of boolean check before attempting. Cosmic might be packaged for a distribution with systemd support, but the end user might not use systemd as their init system (for whatever reason).

In that case, trying and then logging errors is just spamming irrelevant messages for their system.

Comment (skewballfox) [reply src/systemd.rs:41]: plus checking a rwlocked bool is probably cheaper than failing to execute a dbus call, for whatever those few clock cycles are worth

File: src/main.rs
Comment (Drakulix) [reply src/main.rs:422]: > I'm not sure there is a situation where it would be None. I made the value optional for two reasons.

But unwrapping a `None` does panic. We shouldn't assume an api always gives us `Some`, if it's signature says otherwise.

File: src/main.rs
Comment (Drakulix) [reply src/main.rs:431]: Oh wait, that is already what this is doing, I just saw the `unwrap` and got confused, but I guess `spawn_scope` returns a `Result<()>`.

Maybe we can replace the syntax here with something more obvious, like:
```rust
if let Err(err) = spawn_scope(cmd.to_string(), vec![pids]).await {
    warn!("Failed to spawn scope for {}. Creating transient unit failed with {}", cmd, err);
}
```

File: src/main.rs
Comment (skewballfox) [reply src/main.rs]: > However we should make sure to destroy any created scopes, when we detect that the child has exited later. That would at least prevent any process re-using the PID to remain attached to the wrong scope for long.

So I guessing the best path is to map commands to scopes, and assume the pid changed, the scope must be altered. 

So `Dashmap`? `Arc<RwLock<HashMap>>`?

Comment (skewballfox) [reply src/main.rs]: Also, for anyone like me, who was curious. Restarting a failing process is not the responsibility of scopes, at least according to [this doc](https://www.freedesktop.org/software/systemd/man/latest/systemd.scope.html#:~:text=processes%20managed%20as%20scope%20units)

Comment (Drakulix) [reply src/main.rs]: Though this information is local to `cosmic-session`, I kinda dislike, that we already have a list of processes + metadata in the `process_manager` from `launchpad` and are now building our own list for additional metadata (the scope).

Maybe this isn't the best approach. Maybe we could just pass an optional scope to `process_manager.start` and let `launchpad` handle all of this setup. Including destroying the scope, when the process exits.

Alternatively we could allow attaching some arbitrary data to a process launched, which could then be passed to the closures in `with_on_start` and `with_on_exit`.

cc @wash2, any thoughts on this?

Comment (Drakulix) [reply src/main.rs]: > Restarting a failing process is not the responsibility of scopes

Right, it is launch-pads, so our responsibility. So we want to properly cleanup, before restarting.

Comment (skewballfox) [reply src/main.rs]: > Maybe this isn't the best approach. Maybe we could just pass an optional scope to `process_manager.start` and let `launchpad` handle all of this setup. Including destroying the scope, when the process exits.
>
> Alternatively we could allow attaching some arbitrary data to a process launched, which could then be passed to the closures in `with_on_start` and `with_on_exit`. Then we also don't need to depend (even if optional) on systemd in launch-pad.

I'm fine with either. I can recreate the zbus interface stuff in launch-pad. 

It's probably best to keep the option open in either case to have scopes capable of handling more than one PID, and whether it's here or in `launch-pad` there are a lot more capabilities of scopes we might want to make use of, such as cpu or memory limit, or things like on oom, which get will likely be triggered quite a bit if vscode + chrome + cargo or some other build system are active at the same time. maybe a `ScopeConfig` struct?

Comment (Drakulix) [reply src/main.rs]: I don't know, if we actually want to force these values via the programmable interface.

If they have stable names, these options can be set via drop-in config files, which is more flexible to customize for users and/or distros.

File: src/systemd.rs
Comment (skewballfox) [reply src/systemd.rs:41]: Worked on this a little bit. Probably bit twiddling but I'm having trouble doing this without using async, or calling a blocking variant(which zbus docs say may deadlock a thread). Does zbus provide a way of checking for existence of an interface without connecting? I can make the function async but `if *systemd_is_used.await` just kills me a little inside.

File: src/main.rs
Comment (wash2) [reply src/main.rs]: > Alternatively we could allow attaching some arbitrary data to a process launched, which could then be passed to the closures in with_on_start and with_on_exit.
Then we also don't need to depend (even if optional) on systemd in launch-pad.

Ya, if most of the features are better configured through config files, this might be better for now.

Comment (skewballfox) [reply src/main.rs]: Currently trying to move systemd scope spawning to `with_start`, and I'm running into issues moving a clone of command. Process Manager currently doesn't expose a function to get the name of a command or a getter for processdata. It also seems I can't access private methods from within the closure so I can't just directly access inner. In order to pass cmd, I need to capture the evironment (`async move ||`) but I get the error/warning async closures are currently unstable. I'm guessing that switching to unstable is probably not the desired solution.  
<details>
```rust
.with_on_start(move |pman, pkey, will_restart| {
	#[cfg(feature = "systemd")]
	async move || {
		if *is_systemd_used() {
			//currently pid is optional hence the double unwrap
			if let Some(pids) = pman.get_pid(pkey).await.unwrap_or_else(|err| {
				warn!(
					"Failed to get pid for {}. Process manager failed with {}",
					cmd_clone2.clone(),
					err
				);
				None
			}) {
				//spawn_scope takes a vec of pids in case we want to spawn a scope for multiple processes
				if let Err(err) = spawn_scope(cmd_clone2.clone(), vec![pids]).await
				{
					warn!(
		"Failed to spawn scope for {}. Creating transient unit failed with {}",
		cmd_clone2.clone(), err
	);
				};
			}
		}
	}
})
```
</details>

I could open a PR to expose cmd name in with start, but if the problem is moving state, then how would deserializing scope configuration work without moving the systemd stuff to launch-pad? 

slightly unrelated, but is lauch-pad used by anything else, such as pop/cosmic-launcher?

Comment (wash2) [reply src/main.rs]: > slightly unrelated, but is lauch-pad used by anything else, such as pop/cosmic-launcher?

Ya, it is used by cosmic-panel for spawning applets.

> In order to pass cmd, I need to capture the evironment (async move ||) but I get the error/warning async closures are currently unstable. I'm guessing that switching to unstable is probably not the desired solution.

Why do you need `async move || {...}`? Could `move || async move {...}`, or something like that, work for your purposes? `start_component` moves the logging spans into the closure, but you might want to call `to_string` so that you don't run into lifetime issues so that you could do it the same way. Or you could make the argument `cmd: impl Into<Cow<'static, str>>` and convert it to `Cow<'static, str>` right away, so it can be moved into the closures just like the spans. Hopefully I've not misunderstood the issue you are facing here :sweat_smile: 

That being said, I also wouldn't be against a PR that exposes the name from launchpad.

Comment (skewballfox) [reply src/main.rs]: I did an org wide search for launch-pad to see what was using it. Right now it's used in `cosmic-panel`

> > slightly unrelated, but is lauch-pad used by anything else, such as pop/cosmic-launcher?
> 
> Ya, it is used by cosmic-panel for spawning applets.
> 
> > In order to pass cmd, I need to capture the evironment (async move ||) but I get the error/warning async closures are currently unstable. I'm guessing that switching to unstable is probably not the desired solution.
> 
> Why do you need `async move || {...}`? 

I mainly need some way to pass in the command, which I'm using to generate the scope name. so long as we are using a scope per process, that's really the only extra state we need and we can get away with just adding a function call launch-pad side. 

> Could `move || async move {...}`, or something like that, work for your purposes? 

I was not even aware that was a thing, but I tried that (along with the change of cmd to the mentioned typed signature, and I get 
```
`{closure@src/main.rs:384:7: 384:14}` is not a future
the trait `futures_util::Future` is not implemented for closure `{closure@src/main.rs:384:7: 384:14}`
main.rs(384, 7): consider calling this closure
process.rs(113, 6): required by a bound in `launch_pad::process::Process::with_on_start`
```
might be possible to change the function signature of the arg to startwith, but I'm not sure if that's going to break anything.

> That being said, I also wouldn't be against a PR that exposes the name from launchpad.

So I tried using the `get_exe_and_pid` function I added and I'm still having issues related to incompatible types. It's saying sync is not implemented for `dyn futures_util::Future<Output = Result<conn::handshake::Authenticated, zbus::Error>> + std::marker::Send` which (after inlining my function) I think is coming from `zbus::Connection::session()`. 

I'll poke around with this more tomorrow

Comment (wash2) [reply src/main.rs]: Ok, feel free to push what you have for it and ping if you'd like me to take a look too :+1:

Comment (skewballfox) [reply src/main.rs]: @wash2 yeah, kind of stuck on this one. Within the function, I can comment out all the code except the call to `Connection::session()` and the compiler message is still present, so this might be something that is currently missing from zbus. I'm uploading the full compiler diagnostic if it helps
[diagnostics.txt](https://github.com/user-attachments/files/15991367/diagnostics.txt)

Comment (skewballfox) [reply src/main.rs]: Wow, I guess I should learn to read diagnostics better. It turns out that this was adding the feature tokio to zbus

Comment (wash2) [reply src/main.rs]: Ok, thanks :slightly_smiling_face:

Comment (skewballfox) [reply src/main.rs]: hey when you get back around to this. I got some helpful information from the zbus maintainers ([see zbus#869](https://github.com/dbus2/zbus/issues/869)). It turns out it's the future returned by connection (and it seems any awaitable Result). 


I've managed to work around the issue by using `tokio::task::spawn_local`, I still need to test the changes though

Comment (wash2) [reply src/main.rs]: https://github.com/pop-os/launch-pad/pull/14 should fix the issue you are facing, or at least did when I tested it earlier.

File: src/systemd.rs
Comment (skewballfox) [reply src/systemd.rs:41]: @Drakulix so systemctl will throw an error if systemd is not the init system([relevant SO question](https://askubuntu.com/questions/1379425/system-has-not-been-booted-with-systemd-as-init-system-pid-1-cant-operate), do you think it's safe to assume that if `systemctl -v` errors out for any reason, that we should assume systemd is not used

Comment (skewballfox) [reply src/systemd.rs:44]: so I switched to a different method, which it seems is used by [libsystemd](https://manpages.ubuntu.com/manpages/xenial/en/man3/sd_booted.3.html#NOTES:~:text=Internally%2C%20this%20function%20checks%20whether). If you would rather use another method, I have a few other things I could try (like filtering `cat /proc/1/stat`)

File: src/main.rs
Comment (skewballfox) [reply src/main.rs]: It did, I tested it last night. I guess once that is merged, I can go ahead and resolve this.

Comment (wash2) [reply src/main.rs]: @skewballfox I merged https://github.com/pop-os/launch-pad/pull/14, so the Sync bound should be resolved now by updating launch-pad

**Comment by skewballfox** [comment]
What's left to do before merging?

I'm not sure configuring customizable scopes should be part of this PR. In part because almost all of the settable properties for scopes are currently useless (either deprecated or not relevant to cgroup v2 which seems to be the default), and in part because how customization should happen probably warrants some discussion, but just putting each process into it's own cgroup with just defualts would provide immediate benefit to systemd-based systems

**Comment by wash2** [comment]
> What's left to do before merging?

It looks pretty good to me. You could squash some of the commits, but otherwise, I'll just do a bit more testing, and then approve :)

**Comment by skewballfox** [comment]
give me a few, due to the merges with main a few commits aren't mine, and I'm trying to avoid squashing their work into mine or vice versa

File: src/systemd.rs
Comment (Drakulix) [reply src/systemd.rs:44]: seems good to me.

**Review by wash2 — APPROVED** [review]

**Comment by wash2** [comment]
Thanks!

**Pull request merged**
